src/vm/eventpipe.cpp

```diff
@@ -482,22 +482,34 @@ INT_PTR QCALLTYPE EventPipeInternal::CreateProvider(
     return reinterpret_cast<INT_PTR>(pProvider);
 }
 
-INT_PTR QCALLTYPE EventPipeInternal::AddEvent(
+INT_PTR QCALLTYPE EventPipeInternal::DefineEvent(
     INT_PTR provHandle,
-    __int64 keywords,
     unsigned int eventID,
+    __int64 keywords,
     unsigned int eventVersion,
     unsigned int level,
-    bool needStack)
+    void *pMetadata,
+    unsigned int metadataLength)
 {
     QCALL_CONTRACT;
+
+    EventPipeEvent *pEvent = NULL;
+
     BEGIN_QCALL;
 
-    // TODO
+    _ASSERTE(provHandle != NULL);
+    _ASSERTE(pMetadata != NULL);
+    _ASSERTE(metadataLength >= 20);
+    EventPipeProvider *pProvider = reinterpret_cast<EventPipeProvider *>(provHandle);
+    BYTE *buffer = new BYTE[metadataLength];
```

Comment:
Instead of allocating the buffer here, you should allocate it in the EventPipeEvent ctor if an input metadata buffer is passed in.  The metadata buffer needs to live for the life of the EventPipeEvent which is longer than this call.  Then, you can free the buffer in ~EventPipeEvent.